### PR TITLE
Handle Timeout.timeout in rails queries

### PR DESCRIPTION
**Pull request by kbrock**
### TL;DR

If a `Timeout.timeout` is triggered in  `Transaction#within_new_transaction`, an exception is thrown:

```
[NoMethodError]: undefined method `state' for nil:NilClass
  Method:[rescue in block in refresh]
```

### Reproducing Script

[[Script]](https://gist.github.com/kbrock/175a3c7044e3e1f3d6079bd14e6c3cdd) reproduces on rails `5.0.0.1`, `master` with `sqlite3` and `pg`. (did not try mysql)

Since it is a race condition, you sometimes need to run the reproduction script a few times.
Also, I added a `puts "ISSUE" if transaction.nil?` in the code to help me know that I indeed triggered the race condition and that the fix worked.

### Details

When a `SIGINT`, `SIGTERM`, or `Timeout::Error` is thrown inside `within_new_transaction`,
`transaction.nil? == true` and `@stack.empty? == true`

For this reason, the code is currently guarding for `transaction.nil?` in 2 places.

Unfortunately, `commit_transaction` is not guarding against these seemingly impossible cases.

The problem with the method is finally realized in the error checking itself.
it calls `transaction.state` (i.e.: `nil.state`) and that is the final exception
raised.

The solution is to handle `transaction.nil?` and `@stack.empty?`

### Diagram of Details

```ruby
      def within_new_transaction(options = {})
        transaction = begin_transaction options # transaction = nil
        yield
      rescue Exception => error # no exception
        if transaction
          rollback_transaction
          after_failure_actions(transaction, error)
        end
        raise
      ensure
        unless error # no exception
          if Thread.current.status == "aborting" # not aborting
            rollback_transaction if transaction
          else
            begin
              # suggestion: commit_transaction if transaction
              commit_transaction # error
            rescue Exception
              rollback_transaction(transaction) unless transaction.state.completed? # error: transaction == nil
              raise
            end
          end
        end
      end
```

### CODE REVIEW

**Comment by al2o3cr** [comment]
I'd be curious where/why `begin_transaction` is coming up with a `nil` - the [default implementation](https://github.com/rails/rails/blob/92703a9ea5d8b96f30e0b706b801c9185ef14f0e/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb#L153) can't do that under any circumstances...

**Comment by kbrock** [comment]
@al2o3cr I agree with you.

Maybe https://github.com/rails/rails/pull/23407 sheds some light for us.

This exception is masking the sporadic underlying problem. So I can't give you the actual problem for my case.

The PR is based upon the fact that 2 of the 3 transaction methods are protected.
And yes, it is entirely possible that https://github.com/rails/rails/pull/16435 should have actually removed the `if transaction` check when that helper method was added

**Comment by kbrock** [comment]
@al2o3cr Well, `commit_transaction` is throwing the exception because `@stack.last == nil`.

There are a few things blowing up in this code.

I thought it simplest to protect the `commit_transaction` like it is done 2 other places in the same method.

**Comment by Fryguy** [comment]
@kbrock Perhaps then the fix is to use the current_transaction method  in commit_transaction which protects against @stack.last being nil.

**Comment by kbrock** [comment]
Thanks @Fryguy I'll put in that fix.

The null object pattern does suggest we want to go that route instead of if checks everywhere

**Comment by kbrock** [comment]
Thank you @al2o3cr for keeping me honest.

I updated the description with a [[reproduction script]](https://gist.github.com/kbrock/175a3c7044e3e1f3d6079bd14e6c3cdd)

Nice to be asked to verify that a hunch about a problem in production is actually the problem you suspect.

**Comment by kbrock** [comment]
The first commit is a simpler approach and the second commit uses the `current_transaction` null object pattern.

**Comment by kbrock** [comment]
@al2o3cr I updated the description with more details. Does this answer your question?

**Comment by jrafanie** [comment]
@kbrock looks like this PR is conflicted, can you rebase it?

**Comment by kbrock** [comment]
@jrafanie Thanks. Fixed conflict.

The same reproducing script should work (mentioned in the top description

**Comment by kbrock** [comment]
- Rebased to current master.
- Threw away null object pattern complexity.
- This follows the `if transaction` code that is used throughout the code.

@matthewd ideas on how to make this easier for you to merge? (besides more hours in the day)
